src/coreclr/src/vm/interoplibinterface.cpp

```diff
@@ -0,0 +1,1249 @@
+    {
+        NOTHROW;
+        MODE_ANY;
+        PRECONDITION(wrapper != NULL);
+    }
+    CONTRACTL_END;
+
+    InteropLib::Com::DestroyWrapperForObject(wrapper);
+}
+
+void ComWrappersNative::DestroyExternalComObjectContext(_In_ void* contextRaw)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        MODE_ANY;
+        PRECONDITION(contextRaw != NULL);
+    }
+    CONTRACTL_END;
+
+#ifdef _DEBUG
+    ExternalObjectContext* context = static_cast<ExternalObjectContext*>(contextRaw);
+    _ASSERTE(!context->IsActive());
+#endif
+
+    InteropLib::Com::DestroyWrapperForExternal(contextRaw);
+}
+
+void ComWrappersNative::MarkExternalComObjectContextCollected(_In_ void* contextRaw)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        PRECONDITION(contextRaw != NULL);
+        PRECONDITION(GCHeapUtilities::IsGCInProgress());
+    }
+    CONTRACTL_END;
+
+    ExternalObjectContext* context = static_cast<ExternalObjectContext*>(contextRaw);
+    _ASSERTE(context->IsActive());
+    context->MarkCollected();
+
+    // Verify the caller didn't ignore the cache during creation.
+    if (context->IsSet(ExternalObjectContext::Flags_InCache))
+    {
+        ExtObjCxtCache* cache = ExtObjCxtCache::GetInstanceNoThrow();
+        cache->Remove(context);
+    }
+}
+
+#endif // FEATURE_COMINTEROP
+
+void Interop::OnGCStarted(_In_ int nCondemnedGeneration)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+    }
+    CONTRACTL_END;
+
+#ifdef FEATURE_COMINTEROP
+    //
+    // Let GC detect managed/native cycles with input from jupiter
+    // Jupiter will
+    // 1. Report reference from RCW to CCW based on native reference in Jupiter
+    // 2. Identify the subset of CCWs that needs to be rooted
+    //
+    // We'll build the references from RCW to CCW using
+    // 1. Preallocated arrays
+    // 2. Dependent handles
+    //
+    RCWWalker::OnGCStarted(nCondemnedGeneration);
+
+    //
+    // Note that we could get nested GCStart/GCEnd calls, such as :
+    // GCStart for Gen 2 background GC
+    //    GCStart for Gen 0/1 foregorund GC
+    //    GCEnd   for Gen 0/1 foreground GC
+    //    ....
+    // GCEnd for Gen 2 background GC
+    //
+    // The nCondemnedGeneration >= 2 check takes care of this nesting problem
+    //
+    // See Interop::OnGCFinished()
+    if (nCondemnedGeneration >= 2)
+    {
+        EX_TRY
+        {
+            ExtObjCxtCache* cache = ExtObjCxtCache::GetInstance();
+            RCWRefCache *refCache = GetAppDomain()->GetRCWRefCache();
+
+            // Reset the ref cache
+            refCache->ResetDependentHandles();
+
+            // Create a call context for the InteropLib.
+            InteropLibImports::RuntimeCallContext cxt(cache, refCache);
+            (void)InteropLib::Com::BeginExternalObjectReferenceTracking(&cxt);
+
+            // Shrink cache and clear unused handles.
+            refCache->ShrinkDependentHandles();
+        }
+        EX_CATCH
```

Comment:
Throwing and catching exceptions when the runtime is suspect. I expect that you would see a assorted collection of assert/crashes if the exception was actually thrown/caught here.

Reply:
Agreed. I originally didn't have the try/catch. I followed the existing RCW Walking algorithm and it had this try/catch. On top of that the path here indicates a `NOTHROW` environment so I was trying to make it similar. Technically the only throwing part of this code path is `GetAppDomain()->GetRCWRefCache()` since the `GetRCWRefCache()` lazily allocates an instance and can OOM.

I could simply just wrap that call and return if the OOM occurs similar to below:

```C++
if (nCondemnedGeneration >= 2)
{
    RCWRefCache* refCache;
    EX_TRY
    {
        refCache = GetAppDomain()->GetRCWRefCache();
    }
    EX_CATCH
    {
    }
    EX_END_CATCH(RethrowCorruptingExceptions)
    if (refCache == NULL)
        return;
    ExtObjCxtCache* cache = ExtObjCxtCache::GetInstanceNoThrow();

    ...
}
```

Alternatively, I could also handle the OOM entirely in the `GetRCWRefCache()` call and return `NULL`.

Reply:
Why do we need to allocate RCW cache during the GC? If there is no RCW cache, there is nothing to do - no need to allocate one.

Reply:
Not the cache, the ref cache is the issue. The ref cache is the mechanism that associates them - see https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/rcwrefcache.cpp.

Reply:
Just as an FYI, the plan is to update the `RCWRefCache` implementation when we remove the built-in WinRT support. I am simply reusing it here to avoid yet more code.

Reply:
Actually how about I create the `RCWRefCache` at the same time I create the RCW Cache itself. This will allow me to remove the try/catch entirely.

Reply:
Yes, that would be nice.